<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.25">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Robin Waples, Luis Mijangos &amp; Bernd Gruber">
<meta name="description" content="W07 Effective Population Size">

<title>7&nbsp; Effective Population Size – Analysing Genomic Data with dartRverse</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../">
<link href="../../../inst/tutorials/W08/W08.html" rel="next">
<link href="../../../inst/tutorials/W06/W06.html" rel="prev">
<script src="../../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting-7b89279ff1a6dce999919e0e67d4d9ec.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../site_libs/bootstrap/bootstrap-70a47bd5681a7291082a5b9f83d58762.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN" && texText && texText.data) {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../../../dartR_style.css">
</head>

<body class="nav-sidebar floating quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../../inst/tutorials/W07/W07.html"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Effective Population Size</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="../../../index.html" class="sidebar-logo-link">
      </a>
    <div class="sidebar-title mb-0 py-0">
      <a href="../../../">Analysing Genomic Data with dartRverse</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome!</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../schedule.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Program</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../inst/tutorials/W01/W01.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">W01 Pop Gen in Conservation and Restoration</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../inst/tutorials/W02/W02.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">W02 dartR Intro</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../inst/tutorials/W03/W03.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">W03 Sex-Linked Markers</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../inst/tutorials/W04/W04.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">W04 Genetic Structure</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../inst/tutorials/W05/W05.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">W05 Calling SNPs</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../inst/tutorials/W06/W06.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">W06 Small Populations</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../inst/tutorials/W07/W07.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Effective Population Size</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../inst/tutorials/W08/W08.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">W08 Divergence &amp; Phylogenetics</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../inst/tutorials/W09/W09.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">W09 dartR developer</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../inst/tutorials/W10/W10.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">W10 Natural Selection &amp; Adaptation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../inst/tutorials/W11/W11.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">W11 Landscape Genetics</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../inst/tutorials/W12/W12.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">W12 AI in PopGen &amp; Research</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../inst/tutorials/W13/W13.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">W13 Simulations</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../inst/tutorials/W14/W14.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">W14 SNP panels</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../inst/tutorials/W15/W15.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">W15 Relatedness</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../inst/tutorials/W16/W16.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">W16 Sequencing Technologies</span></span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#w07-effective-population-size" id="toc-w07-effective-population-size" class="nav-link active" data-scroll-target="#w07-effective-population-size">W07 Effective Population Size</a></li>
  <li><a href="#effective-population-size" id="toc-effective-population-size" class="nav-link" data-scroll-target="#effective-population-size">Effective Population Size</a></li>
  <li><a href="#learning-outcomes" id="toc-learning-outcomes" class="nav-link" data-scroll-target="#learning-outcomes">Learning Outcomes</a></li>
  <li><a href="#introduction-what-is-effective-population-size" id="toc-introduction-what-is-effective-population-size" class="nav-link" data-scroll-target="#introduction-what-is-effective-population-size">Introduction: What is Effective Population Size?</a>
  <ul>
  <li><a href="#the-concept-of-ne" id="toc-the-concept-of-ne" class="nav-link" data-scroll-target="#the-concept-of-ne">The concept of <em>Ne</em></a></li>
  <li><a href="#why-does-ne-matter-for-conservation" id="toc-why-does-ne-matter-for-conservation" class="nav-link" data-scroll-target="#why-does-ne-matter-for-conservation">Why does <em>Ne</em> matter for conservation?</a></li>
  <li><a href="#the-many-versions-of-ne-which-one-are-you-estimating" id="toc-the-many-versions-of-ne-which-one-are-you-estimating" class="nav-link" data-scroll-target="#the-many-versions-of-ne-which-one-are-you-estimating">The many versions of <em>Ne</em>: which one are you estimating?</a></li>
  </ul></li>
  <li><a href="#part-1-contemporary-ne-the-linkage-disequilibrium-method" id="toc-part-1-contemporary-ne-the-linkage-disequilibrium-method" class="nav-link" data-scroll-target="#part-1-contemporary-ne-the-linkage-disequilibrium-method">Part 1: Contemporary <em>Ne</em> — The Linkage Disequilibrium Method</a>
  <ul>
  <li><a href="#the-ld-approach" id="toc-the-ld-approach" class="nav-link" data-scroll-target="#the-ld-approach">The LD approach</a></li>
  <li><a href="#the-critical-role-of-maf-filtering" id="toc-the-critical-role-of-maf-filtering" class="nav-link" data-scroll-target="#the-critical-role-of-maf-filtering">The critical role of MAF filtering</a></li>
  <li><a href="#running-ld-ne-with-gl.ldne" id="toc-running-ld-ne-with-gl.ldne" class="nav-link" data-scroll-target="#running-ld-ne-with-gl.ldne">Running LD <em>Ne</em> with <code>gl.LDNe</code></a></li>
  <li><a href="#the-waples-correction-for-genomic-data" id="toc-the-waples-correction-for-genomic-data" class="nav-link" data-scroll-target="#the-waples-correction-for-genomic-data">The Waples correction for genomic data</a></li>
  <li><a href="#understanding-inf-estimates" id="toc-understanding-inf-estimates" class="nav-link" data-scroll-target="#understanding-inf-estimates">Understanding <em>Inf</em> estimates</a></li>
  </ul></li>
  <li><a href="#part-2-historical-ne-the-site-frequency-spectrum-approach" id="toc-part-2-historical-ne-the-site-frequency-spectrum-approach" class="nav-link" data-scroll-target="#part-2-historical-ne-the-site-frequency-spectrum-approach">Part 2: Historical <em>Ne</em> — The Site Frequency Spectrum Approach</a>
  <ul>
  <li><a href="#what-is-the-site-frequency-spectrum" id="toc-what-is-the-site-frequency-spectrum" class="nav-link" data-scroll-target="#what-is-the-site-frequency-spectrum">What is the Site Frequency Spectrum?</a></li>
  <li><a href="#computing-the-sfs-in-the-dartrverse" id="toc-computing-the-sfs-in-the-dartrverse" class="nav-link" data-scroll-target="#computing-the-sfs-in-the-dartrverse">2.1 Computing the SFS in the dartRverse</a></li>
  <li><a href="#comparing-sfs-shapes-across-demographic-scenarios" id="toc-comparing-sfs-shapes-across-demographic-scenarios" class="nav-link" data-scroll-target="#comparing-sfs-shapes-across-demographic-scenarios">2.2 Comparing SFS shapes across demographic scenarios</a></li>
  <li><a href="#estimating-historical-ne-with-epos" id="toc-estimating-historical-ne-with-epos" class="nav-link" data-scroll-target="#estimating-historical-ne-with-epos">2.3 Estimating historical <em>Ne</em> with EPOS</a>
  <ul class="collapse">
  <li><a href="#the-key-parameters-l-and-mu" id="toc-the-key-parameters-l-and-mu" class="nav-link" data-scroll-target="#the-key-parameters-l-and-mu">The key parameters: <em>L</em> and <em>mu</em></a></li>
  <li><a href="#what-happens-with-a-wrong-l" id="toc-what-happens-with-a-wrong-l" class="nav-link" data-scroll-target="#what-happens-with-a-wrong-l">What happens with a wrong <em>L</em>?</a></li>
  </ul></li>
  <li><a href="#exercise-4-run-epos-on-all-demographic-scenarios" id="toc-exercise-4-run-epos-on-all-demographic-scenarios" class="nav-link" data-scroll-target="#exercise-4-run-epos-on-all-demographic-scenarios">Exercise 4: Run EPOS on all demographic scenarios</a></li>
  <li><a href="#a-real-data-example" id="toc-a-real-data-example" class="nav-link" data-scroll-target="#a-real-data-example">2.4 A real-data example</a></li>
  <li><a href="#stairways2-an-alternative-coalescent-method" id="toc-stairways2-an-alternative-coalescent-method" class="nav-link" data-scroll-target="#stairways2-an-alternative-coalescent-method">2.5 Stairways2 — an alternative coalescent method</a></li>
  <li><a href="#exercise-5-compare-epos-and-stairways2" id="toc-exercise-5-compare-epos-and-stairways2" class="nav-link" data-scroll-target="#exercise-5-compare-epos-and-stairways2">Exercise 5: Compare EPOS and Stairways2</a></li>
  </ul></li>
  <li><a href="#additional-exercises" id="toc-additional-exercises" class="nav-link" data-scroll-target="#additional-exercises">Additional Exercises</a>
  <ul>
  <li><a href="#exercise-a-apply-the-full-ld-ne-workflow-to-a-real-dataset" id="toc-exercise-a-apply-the-full-ld-ne-workflow-to-a-real-dataset" class="nav-link" data-scroll-target="#exercise-a-apply-the-full-ld-ne-workflow-to-a-real-dataset">Exercise A: Apply the full LD <em>Ne</em> workflow to a real dataset</a></li>
  <li><a href="#exercise-b-estimating-l-from-wattersons-estimator" id="toc-exercise-b-estimating-l-from-wattersons-estimator" class="nav-link" data-scroll-target="#exercise-b-estimating-l-from-wattersons-estimator">Exercise B: Estimating <em>L</em> from Watterson’s estimator</a></li>
  </ul></li>
  <li><a href="#winding-up" id="toc-winding-up" class="nav-link" data-scroll-target="#winding-up">Winding Up</a>
  <ul>
  <li><a href="#discussion-questions" id="toc-discussion-questions" class="nav-link" data-scroll-target="#discussion-questions">Discussion Questions</a></li>
  <li><a href="#where-have-we-come" id="toc-where-have-we-come" class="nav-link" data-scroll-target="#where-have-we-come">Where Have We Come?</a></li>
  </ul></li>
  <li><a href="#further-reading" id="toc-further-reading" class="nav-link" data-scroll-target="#further-reading">Further Reading</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Effective Population Size</span></h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Robin Waples, Luis Mijangos &amp; Bernd Gruber </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">Invalid Date</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="w07-effective-population-size" class="level2">
<h2 class="anchored" data-anchor-id="w07-effective-population-size">W07 Effective Population Size</h2>
<div class="cell">
<div class="cell-output-display">
<script>
  addClassKlippyTo("pre.r, pre.markdown");
  addKlippy('right', 'top', '#000000', '1', 'Copy code', 'Copied!');
</script>
</div>
</div>
</section>
<section id="effective-population-size" class="level2">
<h2 class="anchored" data-anchor-id="effective-population-size">Effective Population Size</h2>
<p><em>Session Presenters</em></p>
<p><img src="images/Presenters5.png" class="img-fluid"></p>
</section>
<section id="learning-outcomes" class="level2">
<h2 class="anchored" data-anchor-id="learning-outcomes">Learning Outcomes</h2>
<p>In this session we will learn how to estimate effective population size (<em>Ne</em>) using genomic SNP data. We start with the theoretical background, then move to two practical approaches: (1) the linkage disequilibrium method for contemporary <em>Ne</em>, and (2) the Site Frequency Spectrum (SFS) / coalescent approach for historical <em>Ne</em>. By the end of this tutorial you should be able to:</p>
<ul>
<li>Explain the concept of <em>Ne</em> and why it differs from census size (<em>Nc</em>)</li>
<li>Distinguish between drift, LD-based, and coalescent <em>Ne</em> estimators</li>
<li>Run a linkage disequilibrium <em>Ne</em> estimate using <code>gl.LDNe</code> in the dartRverse, and understand the role of MAF filtering</li>
<li>Compute a Site Frequency Spectrum (SFS) from SNP data and interpret its shape</li>
<li>Estimate historical <em>Ne</em> trajectories using EPOS, and understand the role of <em>L</em> and <em>mu</em>
<ul>
<li>Appreciate the differences and complementarity of EPOS and Stairways2</li>
</ul></li>
</ul>
<hr>
</section>
<section id="introduction-what-is-effective-population-size" class="level2">
<h2 class="anchored" data-anchor-id="introduction-what-is-effective-population-size">Introduction: What is Effective Population Size?</h2>
<section id="the-concept-of-ne" class="level3">
<h3 class="anchored" data-anchor-id="the-concept-of-ne">The concept of <em>Ne</em></h3>
<p>Effective population size (<em>Ne</em>) is one of the most fundamental concepts in conservation genetics. It represents the size of an <strong>idealised population</strong> that would experience genetic drift, accumulate inbreeding, or generate linkage disequilibrium at the <strong>same rate as the real population under study</strong>.</p>
<p>An idealised population has: equal sex ratio, random mating (panmixia), non-overlapping generations, constant population size over time, and no selection, mutation, or migration.</p>
<p>The actual census size (<em>Nc</em>) — the number of individuals you count in the field — is almost always <strong>larger</strong> than <em>Ne</em>, often dramatically so. The ratio <em>Ne</em>/<em>Nc</em> is typically 0.1–0.5 for vertebrates but can be far smaller in species with highly skewed reproductive success.</p>
<p><strong>Why is <em>Ne</em> &lt; <em>Nc</em>?</strong></p>
<ul>
<li>Unequal sex ratios (e.g.&nbsp;polygynous mating systems where few males do most of the breeding)</li>
<li>High variance in reproductive success (some individuals contribute many offspring, others none)</li>
<li>Population size fluctuations over time (the harmonic mean penalises bottlenecks severely)</li>
<li>Overlapping generations</li>
<li>Population subdivision and restricted gene flow</li>
</ul>
</section>
<section id="why-does-ne-matter-for-conservation" class="level3">
<h3 class="anchored" data-anchor-id="why-does-ne-matter-for-conservation">Why does <em>Ne</em> matter for conservation?</h3>
<p>The rate at which a population loses genetic diversity through drift is governed directly by <em>Ne</em>:</p>
<p><span class="math display">\[H_t = H_0 \left(1 - \frac{1}{2N_e} \right)^t\]</span></p>
<p>where <em>H₀</em> is initial heterozygosity and <em>H_t</em> is heterozygosity after <em>t</em> generations. The practical implications are stark:</p>
<ul>
<li>A population with <em>Ne</em> = 50 loses ~1% of its heterozygosity <strong>per generation</strong></li>
<li>A population with <em>Ne</em> = 500 loses only ~0.1% per generation</li>
<li>A population with <em>Ne</em> = 10 loses nearly 5% per generation — catastrophically fast</li>
</ul>
<p>This is the basis of the <strong>50/500 rule</strong> (Franklin 1980, Soule 1986): a minimum <em>Ne</em> of 50 is needed to avoid severe short-term inbreeding depression, and <em>Ne</em> ≥ 500 (more recently revised upward to ~1000) to maintain long-term evolutionary potential. This is a rule of thumb, not a rigid target — it should be applied alongside species-specific biological knowledge.</p>
<hr>
</section>
<section id="the-many-versions-of-ne-which-one-are-you-estimating" class="level3">
<h3 class="anchored" data-anchor-id="the-many-versions-of-ne-which-one-are-you-estimating">The many versions of <em>Ne</em>: which one are you estimating?</h3>
<p>A critical and often overlooked point is that <strong>different methods estimate different things</strong>. The term “effective population size” covers several conceptually distinct quantities:</p>
<p><strong>Drift <em>Ne</em></strong> is estimated from the change in allele frequencies between two time points (the temporal method). It reflects the rate of allelic drift between samples. Requires sampling the population at two or more time points.</p>
<p><strong>Inbreeding <em>Ne</em></strong> is estimated from pedigrees or genomic excess homozygosity. It reflects the rate at which individuals become more related to each other over time.</p>
<p><strong>Linkage Disequilibrium <em>Ne</em> (LD <em>Ne</em>)</strong> is estimated from the extent of background LD among loci in a single sample. In a finite population, drift creates associations between alleles at unlinked loci — the smaller <em>Ne</em>, the stronger this background LD. This gives a <strong>contemporary</strong> <em>Ne</em>, reflecting the effective size over the past few generations. This is what <code>NeEstimator</code> / <code>gl.LDNe</code> calculates.</p>
<p><strong>Coalescent <em>Ne</em></strong> is estimated from the genealogical history of alleles using the Site Frequency Spectrum (SFS). Different lineages coalesce at rates that depend on <em>Ne</em>, so the distribution of allele frequencies carries a signal of <strong>historical</strong> <em>Ne</em> over hundreds to thousands of generations. EPOS and Stairways2 use this approach.</p>
<blockquote class="blockquote">
<p><strong>Key message:</strong> All <em>Ne</em> estimators compare the observed population to an idealised population, but at different timescales and using different signals. Do not directly compare a contemporary LD <em>Ne</em> to a coalescent <em>Ne</em> without careful consideration — they are measuring related but distinct quantities.</p>
</blockquote>
<hr>
<p>## Setup</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dartRverse)  <span class="co">#version 1.1.2 or higher</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2) <span class="co">#for plotting</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<hr>
</section>
</section>
<section id="part-1-contemporary-ne-the-linkage-disequilibrium-method" class="level2">
<h2 class="anchored" data-anchor-id="part-1-contemporary-ne-the-linkage-disequilibrium-method">Part 1: Contemporary <em>Ne</em> — The Linkage Disequilibrium Method</h2>
<section id="the-ld-approach" class="level3">
<h3 class="anchored" data-anchor-id="the-ld-approach">The LD approach</h3>
<p>The LD method estimates <em>Ne</em> from a <strong>single sample</strong> by measuring background linkage disequilibrium among loci. In an infinitely large population, unlinked loci should be in linkage equilibrium. In a finite population, drift generates LD even among physically unlinked loci. The smaller <em>Ne</em>, the stronger this background LD. The expected squared correlation between alleles at two unlinked loci is approximately (Waples 2006):</p>
<p><span class="math display">\[E[\hat{r}^2] \approx \frac{1}{3 N_e} + \frac{1}{S}\]</span></p>
<p>where <em>S</em> is the sample size. The 1/<em>S</em> term is sampling noise and is added (subtracted) to isolate the drift-induced LD. This is implemented in the <code>NeEstimator</code> software, accessed via <code>gl.LDNe</code> in the dartRverse.</p>
<p><strong>Key assumptions:</strong></p>
<ul>
<li>Loci are in approximate Hardy–Weinberg equilibrium</li>
<li>No strong selection at the measured loci</li>
<li>The sample is representative of a single well-mixed population</li>
</ul>
</section>
<section id="the-critical-role-of-maf-filtering" class="level3">
<h3 class="anchored" data-anchor-id="the-critical-role-of-maf-filtering">The critical role of MAF filtering</h3>
<p>One of the most important quality choices for LD <em>Ne</em> is the <strong>minor allele frequency (MAF) threshold</strong>, set with the <code>critical</code> argument. Rare alleles (very low MAF) are problematic because:</p>
<ol type="1">
<li>Their frequencies are estimated with high uncertainty, especially in small samples</li>
<li>They tend to inflate <span class="math inline">\(\hat{r}^2\)</span> (apparent LD), which biases <em>Ne</em> downward</li>
<li>They may reflect genotyping errors more than true biological signal</li>
</ol>
<p>The standard recommendation is <strong>MAF ≥ 0.05</strong> (<code>critical = 0.05</code>), removing loci where the minor allele occurs in fewer than 5% of gene copies. This balances accuracy (higher threshold is better) against precision (more loci retained means more precise estimates).</p>
<blockquote class="blockquote">
<p>Standard practice: use <code>critical = c(0, 0.05)</code> to compare the unfiltered and MAF-filtered estimates simultaneously. If the two estimates differ markedly, the unfiltered estimate is likely biased by rare alleles.</p>
</blockquote>
</section>
<section id="running-ld-ne-with-gl.ldne" class="level3">
<h3 class="anchored" data-anchor-id="running-ld-ne-with-gl.ldne">Running LD <em>Ne</em> with <code>gl.LDNe</code></h3>
<p>Simulate a population with a known <em>Ne</em> of 50 and estimate it:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># First, download the NeEstimator binary (only necessary once)</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>dir_ne <span class="ot">&lt;-</span> <span class="fu">gl.download.binary</span>(<span class="st">"neestimator"</span>, <span class="at">out.dir =</span> <span class="fu">tempdir</span>())</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Downloaded binary to  C:\Users\s425824\AppData\Local\Temp\Rtmp4akCf0\file85d04ee416 
Unzipped binary to C:\Users\s425824\AppData\Local\Temp\Rtmp4akCf0/neestimator</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulate 50 individuals with 3000 loci</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>sim50 <span class="ot">&lt;-</span> <span class="fu">gl.sim.Neconst</span>(<span class="at">ninds =</span> <span class="dv">50</span>, <span class="at">nlocs =</span> <span class="dv">3000</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Run a few generations of random mating to approach HWE</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>) {</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  sim50 <span class="ot">&lt;-</span> <span class="fu">gl.sim.offspring</span>(sim50, sim50, <span class="at">noffpermother =</span> <span class="dv">1</span>)</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Estimate Ne using LD method</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="co"># critical = c(0, 0.05): once without MAF filter, once with MAF &gt;= 5%</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>ne_result <span class="ot">&lt;-</span> <span class="fu">gl.LDNe</span>(sim50,</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>                     <span class="at">outfile      =</span> <span class="st">"sim50LD.txt"</span>,</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>                     <span class="at">neest.path   =</span> dir_ne,</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>                     <span class="at">critical     =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.05</span>),</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>                     <span class="at">singleton.rm =</span> <span class="cn">TRUE</span>,</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>                     <span class="at">mating       =</span> <span class="st">"random"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.LDNe 
  Processing genlight object with SNP data
Starting gl2genepop 
  Processing genlight object with SNP data
  The genepop file is saved as:  C:\Users\s425824\AppData\Local\Temp\Rtmp4akCf0/dummy.gen/
Completed: gl2genepop 
  Processing genlight object with SNP data</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="W07_files/figure-html/ldne-simulate-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>$offspring_Po
                    Statistic Frequency 1 Frequency 2 Frequency 3
 Lowest Allele Frequency Used       0.050          0+       No S*
    Harmonic Mean Sample Size          50          50          50
      Independent Comparisons     1131760     1919820     1721440
                  OverAll r^2    0.026552    0.025671    0.026087
          Expected r^2 Sample    0.021276    0.021276    0.021276
                Estimated Ne^          61        73.7        67.1
            CI low Parametric        60.2        72.8        66.4
           CI high Parametric        61.9        74.6          68
             CI low JackKnife          52        62.4        56.6
            CI high JackKnife          73        89.1        81.5

  The results are saved in: C:\Users\s425824\AppData\Local\Temp\Rtmp4akCf0/sim50LD.txt 
Completed: gl.LDNe </code></pre>
</div>
</div>
<p>The output shows two rows — one for each <code>critical</code> value. The estimate with <code>critical = 0.05</code> should be closer to the true <em>Ne</em> of 50.</p>
<hr>
<p>### Exercise 1: Effect of MAF threshold on <em>Ne</em> estimates</p>
<div class="callout callout-style-simple callout-none no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">None</span>Exercise: gl.LDNe and MAF filtering
</div>
</div>
<div class="callout-body-container callout-body">
<p>Rerun the example above for different simulated datasets (or your own) Be aware that filtering your dataset is important You want to have high quality data with minimal missingness, and a good number of loci after filtering. Also you would like to have a reasonable sample size (S &gt;= 30) to avoid technical artefacts. Finally maf filtering is critical for accurate estimates, so you should run with a range of thresholds to see how it affects the estimate. For example, you could run with critical = c(0, 0.02, 0.05, 0.10) to see how the estimate changes as you increase the MAF threshold. Definitely remove singletons in most cases.</p>
<p>QUESTIONS: - How does the Ne estimate change as you increase the MAF threshold? - At which threshold does the estimate stabilise? - What are the trade-offs of using very high vs very low MAF thresholds? - What happens if you do not use the HWE simulation step?</p>
</div>
</div>
<div class="cell" data-exercise="true" data-exercise.setup="ex1-setup" data-exercise.timelimit="120">
<div class="gold-panel">
<div class="tutorial-exercise" data-label="ex1" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0" data-pipe="|>">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>dir_ne <span class="ot">&lt;-</span> <span class="fu">gl.download.binary</span>(<span class="st">"neestimator"</span>, <span class="at">out.dir =</span> <span class="fu">tempdir</span>())</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulate 50 individuals with 3000 loci</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>sim50 <span class="ot">&lt;-</span> <span class="fu">gl.sim.Neconst</span>(<span class="at">ninds =</span> <span class="dv">50</span>, <span class="at">nlocs =</span> <span class="dv">3000</span>)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Run a few generations of random mating to approach HWE</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>) {</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>  sim50 <span class="ot">&lt;-</span> <span class="fu">gl.sim.offspring</span>(sim50, sim50, <span class="at">noffpermother =</span> <span class="dv">1</span>)</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Estimate Ne using LD method</span></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a><span class="co"># critical = c(0, 0.05): once without MAF filter, once with MAF &gt;= 5%</span></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>ne_result <span class="ot">&lt;-</span> <span class="fu">gl.LDNe</span>(sim50,</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>                     <span class="at">outfile      =</span> <span class="st">"sim50LD.txt"</span>,</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>                     <span class="at">neest.path   =</span> dir_ne,</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>                     <span class="at">critical     =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.05</span>),</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>                     <span class="at">singleton.rm =</span> <span class="cn">TRUE</span>,</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a>                     <span class="at">mating       =</span> <span class="st">"random"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<script type="application/json" data-ui-opts="1">{"engine":"r","has_checker":false,"caption":"<span data-i18n=\"text.enginecap\" data-i18n-opts=\"{&quot;engine&quot;:&quot;R&quot;}\">R Code<\/span>"}</script>
</div>
</div>
</div>
<hr>
</section>
<section id="the-waples-correction-for-genomic-data" class="level3">
<h3 class="anchored" data-anchor-id="the-waples-correction-for-genomic-data">The Waples correction for genomic data</h3>
<p>When thousands of SNPs are available and a genome map exists, physically linked loci on the <strong>same chromosome</strong> show LD due to recombination — not just drift. This inflates apparent LD and biases <em>Ne</em> downward. Waples et al.&nbsp;(2016) showed that restricting analysis to <strong>inter-chromosomal locus pairs</strong> (<code>pairing = "separate"</code>) corrects for physical linkage and produces improved estimates.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>pops <span class="ot">&lt;-</span> possums.gl[<span class="dv">1</span><span class="sc">:</span><span class="dv">60</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>]</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Assign chromosome information (illustrative — use real data in practice)</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>pops<span class="sc">@</span>chromosome <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(<span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">size =</span> <span class="fu">nLoc</span>(pops), <span class="at">replace =</span> <span class="cn">TRUE</span>))</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>ne_separate <span class="ot">&lt;-</span> <span class="fu">gl.LDNe</span>(pops,</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>                       <span class="at">outfile      =</span> <span class="st">"popsLD_sep.txt"</span>,</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>                       <span class="at">pairing      =</span> <span class="st">"separate"</span>,</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>                       <span class="at">neest.path   =</span> dir_ne,</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>                       <span class="at">critical     =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.05</span>),</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>                       <span class="at">singleton.rm =</span> <span class="cn">TRUE</span>,</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>                       <span class="at">mating       =</span> <span class="st">"random"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.LDNe 
  Processing genlight object with SNP data
Starting gl2genepop 
  Processing genlight object with SNP data
  The genepop file is saved as:  C:\Users\s425824\AppData\Local\Temp\Rtmp4akCf0/dummy.gen/
Completed: gl2genepop 
  Processing genlight object with SNP data</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="W07_files/figure-html/ldne-waples-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>$A
                    Statistic Frequency 1 Frequency 2 Frequency 3
 Lowest Allele Frequency Used       0.050          0+       No S*
    Harmonic Mean Sample Size          30          30          30
      Independent Comparisons        3001        3542        3380
                  OverAll r^2    0.056843    0.055144    0.055995
          Expected r^2 Sample    0.036878    0.036878    0.036878
                Estimated Ne^        13.5          15        14.2
            CI low Parametric        11.4        12.8        12.1
           CI high Parametric          16        17.6        16.7
             CI low JackKnife         8.4          10         9.5
            CI high JackKnife        22.8        23.6        22.3

$B
                    Statistic Frequency 1 Frequency 2 Frequency 3
 Lowest Allele Frequency Used       0.050          0+       No S*
    Harmonic Mean Sample Size          30          30          30
      Independent Comparisons        4125        4471        4125
                  OverAll r^2    0.053429    0.052373    0.053429
          Expected r^2 Sample    0.036878    0.036878    0.036878
                Estimated Ne^        16.7          18        16.7
            CI low Parametric        14.4        15.5        14.4
           CI high Parametric        19.7        21.2        19.7
             CI low JackKnife        11.5        12.2        11.5
            CI high JackKnife        25.8        28.5        25.8

  The results are saved in: C:\Users\s425824\AppData\Local\Temp\Rtmp4akCf0/popsLD_sep.txt 
Completed: gl.LDNe </code></pre>
</div>
</div>
<p>If chromosome data is unavailable but the number of chromosomes (or genome length) is known, apply a numerical correction from Waples et al.&nbsp;(2016):</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>ne_corrected <span class="ot">&lt;-</span> <span class="fu">gl.LDNe</span>(pops,</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>                        <span class="at">outfile                 =</span> <span class="st">"popsLD_corr.txt"</span>,</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">neest.path              =</span> dir_ne,</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">critical                =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.05</span>),</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>                        <span class="at">singleton.rm            =</span> <span class="cn">TRUE</span>,</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>                        <span class="at">mating                  =</span> <span class="st">"random"</span>,</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>                        <span class="at">Waples.correction       =</span> <span class="st">'nChromosomes'</span>,</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>                        <span class="at">Waples.correction.value =</span> <span class="dv">22</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.LDNe 
  Processing genlight object with SNP data
Starting gl2genepop 
  Processing genlight object with SNP data
  The genepop file is saved as:  C:\Users\s425824\AppData\Local\Temp\Rtmp4akCf0/dummy.gen/
Completed: gl2genepop 
  Processing genlight object with SNP data</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="W07_files/figure-html/ldne-correction-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>$A
                            Statistic Frequency 1 Frequency 2 Frequency 3
         Lowest Allele Frequency Used       0.050          0+       No S*
            Harmonic Mean Sample Size          30          30          30
              Independent Comparisons        3321        3916        3741
                          OverAll r^2    0.056847    0.055101    0.055956
                  Expected r^2 Sample    0.036878    0.036878    0.036878
                        Estimated Ne^        13.5          15        14.2
                    CI low Parametric        11.5        12.9        12.2
                   CI high Parametric        15.9        17.5        16.6
                     CI low JackKnife         8.3         9.9         9.4
                    CI high JackKnife          23        23.9        22.6
                 Waples' corrected Ne        17.4        19.4        18.3
  Waples' corrected CI low Parametric        14.8        16.6        15.7
 Waples' corrected CI high Parametric        20.5        22.6        21.4
   Waples' corrected CI low JackKnife        10.7        12.8        12.1
  Waples' corrected CI high JackKnife        29.7        30.8        29.2

$B
                            Statistic Frequency 1 Frequency 2 Frequency 3
         Lowest Allele Frequency Used       0.050          0+       No S*
            Harmonic Mean Sample Size          30          30          30
              Independent Comparisons        4560        4950        4560
                          OverAll r^2    0.054024    0.052873    0.054024
                  Expected r^2 Sample    0.036878    0.036878    0.036878
                        Estimated Ne^        16.1        17.4        16.1
                    CI low Parametric        13.9        15.1        13.9
                   CI high Parametric        18.7        20.2        18.7
                     CI low JackKnife          11        11.7          11
                    CI high JackKnife        24.7        27.6        24.8
                 Waples' corrected Ne        20.8        22.5        20.8
  Waples' corrected CI low Parametric        17.9        19.5        17.9
 Waples' corrected CI high Parametric        24.1        26.1        24.1
   Waples' corrected CI low JackKnife        14.2        15.1        14.2
  Waples' corrected CI high JackKnife        31.9        35.6          32

  The results are saved in: C:\Users\s425824\AppData\Local\Temp\Rtmp4akCf0/popsLD_corr.txt 
Completed: gl.LDNe </code></pre>
</div>
</div>
</section>
<section id="understanding-inf-estimates" class="level3">
<h3 class="anchored" data-anchor-id="understanding-inf-estimates">Understanding <em>Inf</em> estimates</h3>
<p>Sometimes <code>gl.LDNe</code> returns <em>Ne</em> = Inf. This typically means the population is very large (negligible drift-induced LD), but it can also arise as a technical artefact. For random mating with S &gt;= 30 individuals, the bias-corrected formula is:</p>
<p><span class="math display">\[N_e = \frac{1/3 + \sqrt{1/9 - 2.76\hat{r}^{2\prime}}}{2\hat{r}^{2\prime}}\]</span></p>
<p>When the corrected <span class="math inline">\(\hat{r}^{2\prime}\)</span> is very small or negative (which can happen with small samples and rare alleles), the square root becomes undefined and <em>Ne</em> is reported as Inf. The <code>naive = TRUE</code> option skips the bias correction and helps diagnose the source:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>ne_naive <span class="ot">&lt;-</span> <span class="fu">gl.LDNe</span>(pops,</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">outfile      =</span> <span class="st">"popsLD_naive.txt"</span>,</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>                    <span class="at">neest.path   =</span> dir_ne,</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">critical     =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="fl">0.05</span>),</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>                    <span class="at">singleton.rm =</span> <span class="cn">TRUE</span>,</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>                    <span class="at">mating       =</span> <span class="st">"random"</span>,</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>                    <span class="at">naive        =</span> <span class="cn">TRUE</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p><strong>Practical remedies for Inf estimates:</strong></p>
<ul>
<li>Increase the MAF threshold to remove rare, noisy alleles</li>
<li>Check whether sample size is very small (S &lt; 20 is problematic)</li>
<li>If <em>Inf</em> persists across thresholds and with <code>naive = TRUE</code>, it likely reflects a genuinely large population, but be cautious about over-interpreting <em>Inf</em> as “very large” — it may simply mean “too large to estimate with this method and data quality”.</li>
</ul>
<hr>
<div class="callout callout-style-simple callout-none no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">None</span>Exercise 2: Diagnosing <em>Ne</em> with different population sizes
</div>
</div>
<div class="callout-body-container callout-body">
<p>Simulate two populations: - (a) ninds = 200, nlocs = 1000 — a large population - (b) ninds = 20, nlocs = 1000 — a small population - For each, run 3 generations of random mating, then estimate Ne with gl.LDNe. - Use critical = c(0, 0.05).</p>
<p>Questions: - Do you get Inf estimates for the larger population? - Does the MAF filter help? - How close are the estimates to the true Ne?</p>
<p>Please note the estimates are pretty good because those are simulated data sets, hence the assumptions of the method are met. In real data, estimates can be more biased and imprecise due to violations of assumptions, genotyping errors, and other complexities. The key is to understand the limitations and interpret results in context.</p>
</div>
</div>
<div class="cell" data-exercise="true" data-exercise.setup="ex1-setup" data-exercise.timelimit="120">
<div class="gold-panel">
<div class="tutorial-exercise" data-label="ex2" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0" data-pipe="|>">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># EXERCISE:</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulate two populations:</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="co">#   (a) ninds = 200, nlocs = 1000  — a large population</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co">#   (b) ninds = 20,  nlocs = 1000  — a small population</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="co"># For each, run 3 generations of random mating, then estimate Ne with gl.LDNe.</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Use critical = c(0, 0.05).</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="co"># QUESTIONS:</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a><span class="co"># - Do you get Inf estimates for the larger population?</span></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a><span class="co"># - Does the MAF filter help?</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="co"># - How close are the estimates to the true Ne?</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>dir_ne <span class="ot">&lt;-</span> <span class="fu">gl.download.binary</span>(<span class="st">"neestimator"</span>, <span class="at">out.dir =</span> <span class="fu">tempdir</span>())</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Large population (true Ne ~ 200)</span></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>sim200 <span class="ot">&lt;-</span> <span class="fu">gl.sim.Neconst</span>(<span class="at">ninds =</span> <span class="dv">200</span>, <span class="at">nlocs =</span> <span class="dv">1000</span>)</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>) sim200 <span class="ot">&lt;-</span> <span class="fu">gl.sim.offspring</span>(sim200, sim200, <span class="at">noffpermother =</span> <span class="dv">1</span>)</span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Small population (true Ne ~ 20)</span></span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>sim20 <span class="ot">&lt;-</span> <span class="fu">gl.sim.Neconst</span>(<span class="at">ninds =</span> <span class="dv">20</span>, <span class="at">nlocs =</span> <span class="dv">1000</span>)</span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>) sim20 <span class="ot">&lt;-</span> <span class="fu">gl.sim.offspring</span>(sim20, sim20, <span class="at">noffpermother =</span> <span class="dv">1</span>)</span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Run gl.LDNe for both and compare:</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<script type="application/json" data-ui-opts="1">{"engine":"r","has_checker":false,"caption":"<span data-i18n=\"text.enginecap\" data-i18n-opts=\"{&quot;engine&quot;:&quot;R&quot;}\">R Code<\/span>"}</script>
</div>
</div>
</div>
<hr>
</section>
</section>
<section id="part-2-historical-ne-the-site-frequency-spectrum-approach" class="level2">
<h2 class="anchored" data-anchor-id="part-2-historical-ne-the-site-frequency-spectrum-approach">Part 2: Historical <em>Ne</em> — The Site Frequency Spectrum Approach</h2>
<section id="what-is-the-site-frequency-spectrum" class="level3">
<h3 class="anchored" data-anchor-id="what-is-the-site-frequency-spectrum">What is the Site Frequency Spectrum?</h3>
<p>The <strong>Site Frequency Spectrum (SFS)</strong> — also called the allele frequency spectrum — counts how many SNPs in your dataset carry their minor allele at each possible frequency. For a sample of <em>n</em> diploid individuals (2<em>n</em> gene copies), the folded SFS tabulates SNPs from minor allele frequency (MAF) = 1/(2<em>n</em>) up to 0.5.</p>
<p>The shape of the SFS carries a <strong>demographic signal</strong>:</p>
<table class="caption-top table">
<colgroup>
<col style="width: 50%">
<col style="width: 50%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Demographic event</th>
<th style="text-align: left;">Effect on SFS</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Constant population</td>
<td style="text-align: left;">Exponential decay (many rare, few common alleles)</td>
</tr>
<tr class="even">
<td style="text-align: left;">Recent population decline</td>
<td style="text-align: left;">Excess of intermediate-frequency alleles</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Recent population expansion</td>
<td style="text-align: left;">Excess of rare alleles (singletons) relative to intermediate frequencies</td>
</tr>
<tr class="even">
<td style="text-align: left;">Severe bottleneck</td>
<td style="text-align: left;">Strong excess of singletons</td>
</tr>
</tbody>
</table>
<p><strong>Coalescent theory</strong> provides the mathematical link between the SFS and historical <em>Ne</em>: the rate at which lineages coalesce in the past depends on <em>Ne</em> at that time. By fitting a model to the observed SFS, we can infer how <em>Ne</em> has changed over time. This is what EPOS and Stairways2 do.</p>
</section>
<section id="computing-the-sfs-in-the-dartrverse" class="level3">
<h3 class="anchored" data-anchor-id="computing-the-sfs-in-the-dartrverse">2.1 Computing the SFS in the dartRverse</h3>
<p>The <code>gl.sfs</code> function computes the SFS from a genlight/dartR object. Before computing the SFS, filter your data carefully:</p>
<ul>
<li>Minimal missing data (the SFS is sensitive to missingness, which distorts allele frequencies)</li>
<li>Monomorphic loci removed</li>
<li>Good call rates per locus and per individual</li>
</ul>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Example filtering workflow for real data</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>north  <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path.data, <span class="st">"TympoNorth.rds"</span>))</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>north2 <span class="ot">&lt;-</span> <span class="fu">gl.filter.callrate</span>(north,  <span class="at">method =</span> <span class="st">"loc"</span>, <span class="at">threshold =</span> <span class="fl">0.95</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Warning: data include loci that are scored NA across all individuals.
  Consider filtering using gl &lt;- gl.filter.allna(gl)
  Warning: Data may include monomorphic loci in call rate 
                    calculations for filtering
  Recalculating Call Rate
  Removing loci based on Call Rate, threshold = 0.95 </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="W07_files/figure-html/sfs-filters-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.callrate </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>north3 <span class="ot">&lt;-</span> <span class="fu">gl.filter.rdepth</span>(north2,   <span class="at">lower =</span> <span class="dv">10</span>, <span class="at">upper =</span> <span class="dv">40</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.filter.rdepth 
  Processing genlight object with SNP data
  Removing loci with rdepth &lt;= 10 and &gt;= 40 </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="W07_files/figure-html/sfs-filters-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: gl.filter.rdepth </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>north4 <span class="ot">&lt;-</span> <span class="fu">gl.filter.monomorphs</span>(north3)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.filter.monomorphs 
  Processing genlight object with SNP data
  Identifying monomorphic loci
  Removing monomorphic loci and loci with all missing 
                       data
Completed: gl.filter.monomorphs </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>north5 <span class="ot">&lt;-</span> <span class="fu">gl.filter.callrate</span>(north4, <span class="at">threshold =</span> <span class="fl">0.9</span>, <span class="at">method =</span> <span class="st">"ind"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.filter.callrate 
  Processing genlight object with SNP data
  Recalculating Call Rate
  Removing individuals based on Call Rate, threshold = 0.9 </code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="W07_files/figure-html/sfs-filters-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Note: Locus metrics not recalculated
  Note: Resultant monomorphic loci not deleted
Completed: gl.filter.callrate </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nInd</span>(north5)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 121</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nLoc</span>(north5)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 10832</code></pre>
</div>
</div>
<p>Now compute the SFS from a simulated dataset:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load a simulated dataset with constant Ne = 100</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>gl100 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path.data, <span class="st">"slim_100.rds"</span>))</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="fu">nInd</span>(gl100)</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="fu">nLoc</span>(gl100)</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Folded SFS (default — use when ancestral allele state is unknown)</span></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>sfs_const <span class="ot">&lt;-</span> <span class="fu">gl.sfs</span>(gl100, <span class="at">folded =</span> <span class="cn">TRUE</span>)</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>sfs_const</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Unfolded SFS (requires knowledge of the ancestral allele)</span></span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a>sfs_unfolded <span class="ot">&lt;-</span> <span class="fu">gl.sfs</span>(gl100, <span class="at">folded =</span> <span class="cn">FALSE</span>)</span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a>sfs_unfolded</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>The <code>minbinsize</code> argument controls whether the lowest-frequency bins are included. <code>minbinsize = 1</code> retains singletons. In real data, singletons can be inflated by sequencing errors, so <code>minbinsize = 2</code> or higher is common for empirical datasets. However, removing too many bins loses demographic information — the lowest bins carry the strongest signal about recent events.</p>
<hr>
<p>### Exercise 3: Exploring the SFS</p>
<div class="cell" data-exercise="true" data-exercise.setup="ex3-setup">
<div class="gold-panel">
<div class="tutorial-exercise" data-label="ex3" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0" data-pipe="|>">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># EXERCISE:</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Using an example data set </span></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Steps:</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="co">#   1. gl.filter.callrate(method = "loc", threshold = 0.95) to filter missing data</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a><span class="co">#   2. gl.filter.monomorphs() to remove invariant loci</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a><span class="co">#   3. gl.sfs(folded = TRUE) with minbinsize = 1 and then minbinsize = 5</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a><span class="co"># QUESTIONS:</span></span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a><span class="co"># - What does the SFS look like? Does it suggest recent expansion or contraction?</span></span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a><span class="co"># - How does changing minbinsize affect the number of SNPs in the spectrum?</span></span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a><span class="co"># - When would you use a higher minbinsize in real data?</span></span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>glex    <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path.data, <span class="st">"slim_ex.rds"</span>))</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<script type="application/json" data-ui-opts="1">{"engine":"r","has_checker":false,"caption":"<span data-i18n=\"text.enginecap\" data-i18n-opts=\"{&quot;engine&quot;:&quot;R&quot;}\">R Code<\/span>"}</script>
</div>
</div>
</div>
<hr>
</section>
<section id="comparing-sfs-shapes-across-demographic-scenarios" class="level3">
<h3 class="anchored" data-anchor-id="comparing-sfs-shapes-across-demographic-scenarios">2.2 Comparing SFS shapes across demographic scenarios</h3>
<p>Let us compare four simulated datasets to build intuition about SFS shapes under different demographic histories.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load four simulated datasets with known demographic histories</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>gl100    <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path.data, <span class="st">"slim_100.rds"</span>))</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>gldec    <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path.data, <span class="st">"slim_100_50_50yago.rds"</span>))</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>glinc    <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path.data, <span class="st">"slim_100_200_50yago.rds"</span>))</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>glbottle <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path.data, <span class="st">"slim_100_10_50_50yago_10year.rds"</span>))</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute SFS for each scenario</span></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>sfs_const  <span class="ot">&lt;-</span> <span class="fu">gl.sfs</span>(gl100,    <span class="at">folded =</span> <span class="cn">TRUE</span>)</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>sfs_dec    <span class="ot">&lt;-</span> <span class="fu">gl.sfs</span>(gldec,    <span class="at">folded =</span> <span class="cn">TRUE</span>)</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>sfs_inc    <span class="ot">&lt;-</span> <span class="fu">gl.sfs</span>(glinc,    <span class="at">folded =</span> <span class="cn">TRUE</span>)</span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>sfs_bottle <span class="ot">&lt;-</span> <span class="fu">gl.sfs</span>(glbottle, <span class="at">folded =</span> <span class="cn">TRUE</span>)</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine into a single data frame and plot</span></span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">x          =</span> <span class="dv">0</span><span class="sc">:</span><span class="dv">50</span>,</span>
<span id="cb33-16"><a href="#cb33-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">Constant   =</span> <span class="fu">as.numeric</span>(sfs_const),</span>
<span id="cb33-17"><a href="#cb33-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">Decline    =</span> <span class="fu">as.numeric</span>(sfs_dec),</span>
<span id="cb33-18"><a href="#cb33-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">Increase   =</span> <span class="fu">as.numeric</span>(sfs_inc),</span>
<span id="cb33-19"><a href="#cb33-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">Bottleneck =</span> <span class="fu">as.numeric</span>(sfs_bottle)</span>
<span id="cb33-20"><a href="#cb33-20" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb33-21"><a href="#cb33-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-22"><a href="#cb33-22" aria-hidden="true" tabindex="-1"></a>df_long <span class="ot">&lt;-</span> tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(df, <span class="sc">-</span>x, <span class="at">names_to =</span> <span class="st">"Scenario"</span>, <span class="at">values_to =</span> <span class="st">"Count"</span>)</span>
<span id="cb33-23"><a href="#cb33-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-24"><a href="#cb33-24" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_long, <span class="fu">aes</span>(<span class="at">x =</span> x, <span class="at">y =</span> Count)) <span class="sc">+</span></span>
<span id="cb33-25"><a href="#cb33-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">fill =</span> <span class="st">"steelblue"</span>) <span class="sc">+</span></span>
<span id="cb33-26"><a href="#cb33-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>Scenario, <span class="at">scales =</span> <span class="st">"free_y"</span>) <span class="sc">+</span></span>
<span id="cb33-27"><a href="#cb33-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x     =</span> <span class="st">"Minor allele frequency class"</span>,</span>
<span id="cb33-28"><a href="#cb33-28" aria-hidden="true" tabindex="-1"></a>       <span class="at">y     =</span> <span class="st">"Number of SNPs"</span>,</span>
<span id="cb33-29"><a href="#cb33-29" aria-hidden="true" tabindex="-1"></a>       <span class="at">title =</span> <span class="st">"SFS shapes under different demographic histories"</span>) <span class="sc">+</span> <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">4000</span>))</span>
<span id="cb33-30"><a href="#cb33-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Once you can see four different sfs within a panel you can clearly see the effect of historic population sizes. Mind you those data sets</p>
<p>Differences between SFS shapes can be subtle visually. This is why formal inference methods are needed — human visual comparison of SFS plots is unreliable, especially for mild or recent demographic events. If you have a real dataset it might be a good idea to simulate data for a certain trajectory using your sample sizes and number of SNPs and then compare this to your sfs.</p>
<hr>
</section>
<section id="estimating-historical-ne-with-epos" class="level3">
<h3 class="anchored" data-anchor-id="estimating-historical-ne-with-epos">2.3 Estimating historical <em>Ne</em> with EPOS</h3>
<p><strong>EPOS</strong> (Lynch et al.&nbsp;2019) estimates the historical <em>Ne</em> trajectory using a fast semi-analytical optimisation applied to the SFS. We use EPOS in this tutorial because it runs in seconds to minutes rather than the hours required by Stairways2, while producing broadly comparable results.</p>
<section id="the-key-parameters-l-and-mu" class="level4">
<h4 class="anchored" data-anchor-id="the-key-parameters-l-and-mu">The key parameters: <em>L</em> and <em>mu</em></h4>
<p>EPOS requires two parameters to place the inferred trajectory on an absolute scale:</p>
<ul>
<li><strong>L</strong> — the total length of the genome region that was sequenced (base pairs). This is not the number of SNPs, but the total sequence length surveyed — the “opportunity” for mutations to occur.</li>
<li><strong>mu</strong> — the per-site per-generation mutation rate.</li>
</ul>
<p>These together determine the expected number of mutations per generation:</p>
<p><span class="math display">\[x = \mu \times L \times 2 \times N_e\]</span></p>
<p>If <em>L</em> or <em>mu</em> is wrong, the <strong>shape</strong> of the trajectory is usually preserved but the <strong>absolute axis values</strong> (Ne and generations) will be off. For within-dataset comparisons, the trajectory shape is more robust. For comparing absolute <em>Ne</em> across datasets or species, getting <em>L</em> and <em>mu</em> right is essential.</p>
<p><strong>For simulated data:</strong> <em>L</em> and <em>mu</em> are known exactly.</p>
<p><strong>For DArTseq data:</strong> A common approximation is <code>L = nLoc * 75 * 200</code> (loci × tag length ~75 bp × ~200 to account for sampling fraction). Alternatively, <em>L</em> can be estimated using Watterson’s formula (see Exercise B). Species-specific mutation rates should be used where available; a vertebrate default is often <code>mu = 1e-8</code>.</p>
<p>First, download EPOS:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>dir_epos <span class="ot">&lt;-</span> <span class="fu">gl.download.binary</span>(<span class="st">"epos"</span>, <span class="at">out.dir =</span> <span class="fu">tempdir</span>())</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>Run EPOS on the constant <em>Ne</em> = 100 simulation:</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>L  <span class="ot">&lt;-</span> <span class="fl">5e8</span>   <span class="co"># total genome length from simulation (known)</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>mu <span class="ot">&lt;-</span> <span class="fl">1e-8</span>  <span class="co"># mutation rate from simulation (known)</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>Ne_epos <span class="ot">&lt;-</span> <span class="fu">gl.epos</span>(gl100,</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>                   <span class="at">epos.path  =</span> path.binaries,</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">l          =</span> L,</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">u          =</span> mu,</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>                   <span class="at">boot       =</span> <span class="dv">10</span>,</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>                   <span class="at">minbinsize =</span> <span class="dv">1</span>)</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Ne_epos) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"generation"</span>, <span class="st">"low"</span>, <span class="st">"Ne_median"</span>, <span class="st">"high"</span>)</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(Ne_epos, <span class="fu">aes</span>(<span class="at">x =</span> generation, <span class="at">y =</span> Ne_median)) <span class="sc">+</span></span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">colour =</span> <span class="st">"steelblue"</span>, <span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_ribbon</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> low, <span class="at">ymax =</span> high), <span class="at">alpha =</span> <span class="fl">0.2</span>, <span class="at">fill =</span> <span class="st">"steelblue"</span>) <span class="sc">+</span></span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x     =</span> <span class="st">"Generations ago"</span>,</span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a>       <span class="at">y     =</span> <span class="st">"Effective population size (Ne)"</span>,</span>
<span id="cb35-18"><a href="#cb35-18" aria-hidden="true" tabindex="-1"></a>       <span class="at">title =</span> <span class="st">"EPOS estimate: constant Ne = 100"</span>) <span class="sc">+</span></span>
<span id="cb35-19"><a href="#cb35-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>The median should hover around <em>Ne</em> = 100 throughout. Confidence intervals widen in the most recent time periods (too few mutations have accumulated for precise estimation) and in the deepest past (the coalescent is poorly resolved by a finite sample).</p>
</section>
<section id="what-happens-with-a-wrong-l" class="level4">
<h4 class="anchored" data-anchor-id="what-happens-with-a-wrong-l">What happens with a wrong <em>L</em>?</h4>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Common but incorrect approximation: nLoc * 69 bp</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>L_wrong <span class="ot">&lt;-</span> <span class="dv">69</span> <span class="sc">*</span> <span class="fu">nLoc</span>(gl100)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>Ne_epos_wrong <span class="ot">&lt;-</span> <span class="fu">gl.epos</span>(gl100,</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>                         <span class="at">epos.path  =</span> path.binaries,</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>                         <span class="at">l          =</span> L_wrong,</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>                         <span class="at">u          =</span> mu,</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>                         <span class="at">boot       =</span> <span class="dv">10</span>,</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>                         <span class="at">minbinsize =</span> <span class="dv">1</span>)</span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Ne_epos_wrong) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"generation"</span>, <span class="st">"low"</span>, <span class="st">"Ne_median"</span>, <span class="st">"high"</span>)</span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(Ne_epos_wrong, <span class="fu">aes</span>(<span class="at">x =</span> generation, <span class="at">y =</span> Ne_median)) <span class="sc">+</span></span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">colour =</span> <span class="st">"tomato"</span>, <span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_ribbon</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> low, <span class="at">ymax =</span> high), <span class="at">alpha =</span> <span class="fl">0.2</span>, <span class="at">fill =</span> <span class="st">"tomato"</span>) <span class="sc">+</span></span>
<span id="cb36-16"><a href="#cb36-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x     =</span> <span class="st">"Generations ago"</span>,</span>
<span id="cb36-17"><a href="#cb36-17" aria-hidden="true" tabindex="-1"></a>       <span class="at">y     =</span> <span class="st">"Ne"</span>,</span>
<span id="cb36-18"><a href="#cb36-18" aria-hidden="true" tabindex="-1"></a>       <span class="at">title =</span> <span class="fu">paste0</span>(<span class="st">"EPOS with L = nLoc x 69 = "</span>, L_wrong, <span class="st">" (incorrect)"</span>)) <span class="sc">+</span></span>
<span id="cb36-19"><a href="#cb36-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p>The trajectory <strong>shape</strong> is preserved, but absolute axis values are badly wrong. Setting <em>L</em> too low inflates the inferred <em>Ne</em> — the same number of observed mutations must be explained by a smaller genome, implying a larger population. Importantly, the relative pattern of increases and decreases remains informative even when <em>L</em> is misspecified.</p>
<hr>
</section>
</section>
<section id="exercise-4-run-epos-on-all-demographic-scenarios" class="level3">
<h3 class="anchored" data-anchor-id="exercise-4-run-epos-on-all-demographic-scenarios">Exercise 4: Run EPOS on all demographic scenarios</h3>
<div class="cell" data-exercise="true" data-exercise.setup="ex4-setup">
<div class="gold-panel">
<div class="tutorial-exercise" data-label="ex4" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0" data-pipe="|>">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># EXERCISE:</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Run EPOS on all four simulated datasets (gl100, glinc, gldec, glbottle).</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Use L = 5e8 and mu = 1e-8 for all four.</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot the four Ne trajectories side by side.</span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a><span class="co"># QUESTIONS:</span></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a><span class="co"># - Can EPOS correctly identify the direction of demographic change?</span></span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a><span class="co"># - How well does it recover the timing of the change?</span></span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a><span class="co"># - Which scenario is hardest to recover? Why?</span></span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a>L  <span class="ot">&lt;-</span> <span class="fl">5e8</span></span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a>mu <span class="ot">&lt;-</span> <span class="fl">1e-8</span></span>
<span id="cb37-13"><a href="#cb37-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-14"><a href="#cb37-14" aria-hidden="true" tabindex="-1"></a>gl100    <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path.data, <span class="st">"slim_100.rds"</span>))</span>
<span id="cb37-15"><a href="#cb37-15" aria-hidden="true" tabindex="-1"></a>gldec    <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path.data, <span class="st">"slim_100_50_50yago.rds"</span>))</span>
<span id="cb37-16"><a href="#cb37-16" aria-hidden="true" tabindex="-1"></a>glinc    <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path.data, <span class="st">"slim_100_200_50yago.rds"</span>))</span>
<span id="cb37-17"><a href="#cb37-17" aria-hidden="true" tabindex="-1"></a>glbottle <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path.data, <span class="st">"slim_100_10_50_50yago_10year.rds"</span>))</span>
<span id="cb37-18"><a href="#cb37-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-19"><a href="#cb37-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Tip: run gl.epos() on each, then use patchwork to plot them together:</span></span>
<span id="cb37-20"><a href="#cb37-20" aria-hidden="true" tabindex="-1"></a><span class="co"># p1 + p2 + p3 + p4</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<script type="application/json" data-ui-opts="1">{"engine":"r","has_checker":false,"caption":"<span data-i18n=\"text.enginecap\" data-i18n-opts=\"{&quot;engine&quot;:&quot;R&quot;}\">R Code<\/span>"}</script>
</div>
</div>
</div>
<hr>
</section>
<section id="a-real-data-example" class="level3">
<h3 class="anchored" data-anchor-id="a-real-data-example">2.4 A real-data example</h3>
<p>Here we apply the full workflow to a real SNP dataset from northern brush-tailed possums (<em>Trichosurus vulpecula</em>, Tympo North population). The main difference from simulated data is that <em>L</em> must be approximated.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load and filter real data</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>north  <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path.data, <span class="st">"TympoNorth.rds"</span>))</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>north2 <span class="ot">&lt;-</span> <span class="fu">gl.filter.callrate</span>(north,  <span class="at">method =</span> <span class="st">"loc"</span>, <span class="at">threshold =</span> <span class="fl">0.95</span>)</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>north3 <span class="ot">&lt;-</span> <span class="fu">gl.filter.rdepth</span>(north2,   <span class="at">lower =</span> <span class="dv">10</span>, <span class="at">upper =</span> <span class="dv">40</span>)</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>north4 <span class="ot">&lt;-</span> <span class="fu">gl.filter.monomorphs</span>(north3)</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>north5 <span class="ot">&lt;-</span> <span class="fu">gl.filter.callrate</span>(north4, <span class="at">threshold =</span> <span class="fl">0.9</span>, <span class="at">method =</span> <span class="st">"ind"</span>)</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a><span class="fu">nInd</span>(north5)</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a><span class="fu">nLoc</span>(north5)</span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a><span class="co"># For DArT data: L ~ nLoc * 75 bp * 200</span></span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a>L_real  <span class="ot">&lt;-</span> <span class="fu">nLoc</span>(north5) <span class="sc">*</span> <span class="dv">75</span> <span class="sc">*</span> <span class="dv">200</span></span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a>mu_real <span class="ot">&lt;-</span> <span class="fl">16.17e-9</span>     <span class="co"># possum-specific estimate</span></span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Run EPOS</span></span>
<span id="cb38-16"><a href="#cb38-16" aria-hidden="true" tabindex="-1"></a>Ne_north <span class="ot">&lt;-</span> <span class="fu">gl.run.epos</span>(north5,</span>
<span id="cb38-17"><a href="#cb38-17" aria-hidden="true" tabindex="-1"></a>                        <span class="at">epos.path  =</span> path.binaries,</span>
<span id="cb38-18"><a href="#cb38-18" aria-hidden="true" tabindex="-1"></a>                        <span class="at">L          =</span> L_real,</span>
<span id="cb38-19"><a href="#cb38-19" aria-hidden="true" tabindex="-1"></a>                        <span class="at">u          =</span> mu_real,</span>
<span id="cb38-20"><a href="#cb38-20" aria-hidden="true" tabindex="-1"></a>                        <span class="at">method     =</span> <span class="st">"greedy"</span>,</span>
<span id="cb38-21"><a href="#cb38-21" aria-hidden="true" tabindex="-1"></a>                        <span class="at">depth      =</span> <span class="dv">2</span>,</span>
<span id="cb38-22"><a href="#cb38-22" aria-hidden="true" tabindex="-1"></a>                        <span class="at">boot       =</span> <span class="dv">5</span>)</span>
<span id="cb38-23"><a href="#cb38-23" aria-hidden="true" tabindex="-1"></a>Ne_north<span class="sc">$</span>plot</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<hr>
</section>
<section id="stairways2-an-alternative-coalescent-method" class="level3">
<h3 class="anchored" data-anchor-id="stairways2-an-alternative-coalescent-method">2.5 Stairways2 — an alternative coalescent method</h3>
<p><strong>Stairways2</strong> (Liu &amp; Fu 2020) is the most widely used method for historical <em>Ne</em> inference from the SFS. It uses a machine learning approach to explore the demographic parameter space more thoroughly than EPOS’s semi-analytical optimisation, which may give an advantage for complex multi-epoch histories.</p>
<p>The key trade-off is <strong>computational time</strong>: a Stairways2 run that takes 5–60 minutes can be accomplished in seconds with EPOS, and results are often broadly similar. Our recommended workflow is:</p>
<ol type="1">
<li>Use <strong>EPOS</strong> for rapid exploration and quality checking</li>
<li>Use <strong>Stairways2</strong> for final publishable results when time allows</li>
</ol>
<p>Both methods use the <strong>same inputs</strong> (<em>L</em>, <em>mu</em>, <code>minbinsize</code>), so the discussion above applies equally to both, and results are directly comparable.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Stairways2 is slow (5+ minutes) — run and save, then load results</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Un-comment to actually run:</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Ne_sw &lt;- gl.stairway2(gl100,</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="co">#                       stairway.path = path.binaries,</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a><span class="co">#                       mu            = 1e-8,</span></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a><span class="co">#                       gentime       = 1,</span></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a><span class="co">#                       run           = TRUE,</span></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a><span class="co">#                       nreps         = 30,</span></span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a><span class="co">#                       parallel      = 4,</span></span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a><span class="co">#                       L             = 5e8,</span></span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a><span class="co">#                       minbinsize    = 1)</span></span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a><span class="co"># saveRDS(Ne_sw, file.path(path.data, "Ne_sw_gl100.rds"))</span></span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Load pre-computed result</span></span>
<span id="cb39-15"><a href="#cb39-15" aria-hidden="true" tabindex="-1"></a>Ne_sw <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path.data, <span class="st">"Ne_sw_gl100.rds"</span>))</span>
<span id="cb39-16"><a href="#cb39-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-17"><a href="#cb39-17" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(Ne_sw, <span class="fu">aes</span>(<span class="at">x =</span> year, <span class="at">y =</span> Ne_median)) <span class="sc">+</span></span>
<span id="cb39-18"><a href="#cb39-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">colour =</span> <span class="st">"steelblue"</span>, <span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb39-19"><a href="#cb39-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_ribbon</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> Ne_2.<span class="fl">5.</span>, <span class="at">ymax =</span> Ne_97.<span class="fl">5.</span>),</span>
<span id="cb39-20"><a href="#cb39-20" aria-hidden="true" tabindex="-1"></a>              <span class="at">alpha =</span> <span class="fl">0.2</span>, <span class="at">fill =</span> <span class="st">"steelblue"</span>) <span class="sc">+</span></span>
<span id="cb39-21"><a href="#cb39-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x     =</span> <span class="st">"Years ago"</span>,</span>
<span id="cb39-22"><a href="#cb39-22" aria-hidden="true" tabindex="-1"></a>       <span class="at">y     =</span> <span class="st">"Effective population size (Ne)"</span>,</span>
<span id="cb39-23"><a href="#cb39-23" aria-hidden="true" tabindex="-1"></a>       <span class="at">title =</span> <span class="st">"Stairways2: constant Ne = 100"</span>) <span class="sc">+</span></span>
<span id="cb39-24"><a href="#cb39-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<hr>
</section>
<section id="exercise-5-compare-epos-and-stairways2" class="level3">
<h3 class="anchored" data-anchor-id="exercise-5-compare-epos-and-stairways2">Exercise 5: Compare EPOS and Stairways2</h3>
<div class="cell" data-exercise="true" data-exercise.setup="ex4-setup">
<div class="gold-panel">
<div class="tutorial-exercise" data-label="ex5" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0" data-pipe="|>">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># EXERCISE:</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the pre-computed Stairways2 result:</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="co">#   Ne_sw &lt;- readRDS(file.path(path.data, "Ne_sw_gl100.rds"))</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Run EPOS on the same dataset (gl100, L = 5e8, mu = 1e-8).</span></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot both trajectories on the same graph for direct comparison.</span></span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a><span class="co"># HINT: Stairways2 returns "year" on the x-axis; EPOS returns "generation".</span></span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a><span class="co"># If generation time = 1 year they are equivalent.</span></span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a><span class="co"># QUESTIONS:</span></span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a><span class="co"># - Do EPOS and Stairways2 agree on the general trajectory?</span></span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a><span class="co"># - Do the confidence intervals differ in width?</span></span>
<span id="cb40-14"><a href="#cb40-14" aria-hidden="true" tabindex="-1"></a><span class="co"># - In what circumstances would you prefer Stairways2 over EPOS?</span></span>
<span id="cb40-15"><a href="#cb40-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-16"><a href="#cb40-16" aria-hidden="true" tabindex="-1"></a>gl100 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path.data, <span class="st">"slim_100.rds"</span>))</span>
<span id="cb40-17"><a href="#cb40-17" aria-hidden="true" tabindex="-1"></a>Ne_sw <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">file.path</span>(path.data, <span class="st">"Ne_sw_gl100.rds"</span>))</span>
<span id="cb40-18"><a href="#cb40-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-19"><a href="#cb40-19" aria-hidden="true" tabindex="-1"></a>L  <span class="ot">&lt;-</span> <span class="fl">5e8</span></span>
<span id="cb40-20"><a href="#cb40-20" aria-hidden="true" tabindex="-1"></a>mu <span class="ot">&lt;-</span> <span class="fl">1e-8</span></span>
<span id="cb40-21"><a href="#cb40-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-22"><a href="#cb40-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Your code here:</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<script type="application/json" data-ui-opts="1">{"engine":"r","has_checker":false,"caption":"<span data-i18n=\"text.enginecap\" data-i18n-opts=\"{&quot;engine&quot;:&quot;R&quot;}\">R Code<\/span>"}</script>
</div>
</div>
</div>
<hr>
</section>
</section>
<section id="additional-exercises" class="level2">
<h2 class="anchored" data-anchor-id="additional-exercises">Additional Exercises</h2>
<section id="exercise-a-apply-the-full-ld-ne-workflow-to-a-real-dataset" class="level3">
<h3 class="anchored" data-anchor-id="exercise-a-apply-the-full-ld-ne-workflow-to-a-real-dataset">Exercise A: Apply the full LD <em>Ne</em> workflow to a real dataset</h3>
<div class="cell" data-exercise="true" data-exercise.setup="exA-setup">
<div class="gold-panel">
<div class="tutorial-exercise" data-label="exA" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0" data-pipe="|>">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># EXERCISE:</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Using possums.gl, choose two populations of contrasting size or history.</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Run the full LD Ne workflow:</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a><span class="co">#   1. gl.keep.pop() to keep each population separately</span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a><span class="co">#   2. gl.LDNe() with critical = c(0, 0.05)</span></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare the Ne estimates for the two populations.</span></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a><span class="co"># QUESTIONS:</span></span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a><span class="co"># - Which population has the lower Ne?</span></span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a><span class="co"># - What does a low Ne imply for conservation management of that population?</span></span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a><span class="co"># - Do the Ne estimates change notably between critical = 0 and critical = 0.05?</span></span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>dir_ne <span class="ot">&lt;-</span> <span class="fu">gl.download.binary</span>(<span class="st">"neestimator"</span>, <span class="at">out.dir =</span> <span class="fu">tempdir</span>())</span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Your code here:</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<script type="application/json" data-ui-opts="1">{"engine":"r","has_checker":false,"caption":"<span data-i18n=\"text.enginecap\" data-i18n-opts=\"{&quot;engine&quot;:&quot;R&quot;}\">R Code<\/span>"}</script>
</div>
</div>
</div>
<hr>
</section>
<section id="exercise-b-estimating-l-from-wattersons-estimator" class="level3">
<h3 class="anchored" data-anchor-id="exercise-b-estimating-l-from-wattersons-estimator">Exercise B: Estimating <em>L</em> from Watterson’s estimator</h3>
<div class="cell" data-exercise="true" data-exercise.setup="exB-setup">
<div class="gold-panel">
<div class="tutorial-exercise" data-label="exB" data-completion="1" data-diagnostics="1" data-startover="1" data-lines="0" data-pipe="|>">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># EXERCISE:</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="co"># When L is unknown, it can be estimated from Watterson's theta.</span></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a><span class="co"># The formula is:</span></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a><span class="co">#   L = sum(sfs[-1]) / (4 * n * mu * a_n)</span></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a><span class="co"># where a_n = sum(1/i for i = 1 to 2n-1) is the Watterson normalisation constant,</span></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a><span class="co"># n is the number of individuals, and mu is the mutation rate.</span></span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Use this formula to estimate L for the sim50 dataset (true L = 5e8),</span></span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a><span class="co"># then run EPOS with L_est and compare to the result with the true L.</span></span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>sim50 <span class="ot">&lt;-</span> <span class="fu">gl.sim.Neconst</span>(<span class="at">ninds =</span> <span class="dv">50</span>, <span class="at">nlocs =</span> <span class="dv">3000</span>)</span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>) sim50 <span class="ot">&lt;-</span> <span class="fu">gl.sim.offspring</span>(sim50, sim50, <span class="at">noffpermother =</span> <span class="dv">1</span>)</span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a>mu <span class="ot">&lt;-</span> <span class="fl">1e-8</span></span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a>n  <span class="ot">&lt;-</span> <span class="fu">nInd</span>(sim50)</span>
<span id="cb42-16"><a href="#cb42-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-17"><a href="#cb42-17" aria-hidden="true" tabindex="-1"></a>sfs <span class="ot">&lt;-</span> <span class="fu">gl.sfs</span>(sim50)</span>
<span id="cb42-18"><a href="#cb42-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-19"><a href="#cb42-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Watterson normalisation constant</span></span>
<span id="cb42-20"><a href="#cb42-20" aria-hidden="true" tabindex="-1"></a>a_n <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="dv">1</span> <span class="sc">/</span> (<span class="dv">1</span><span class="sc">:</span>(<span class="dv">2</span><span class="sc">*</span>n <span class="sc">-</span> <span class="dv">1</span>)))</span>
<span id="cb42-21"><a href="#cb42-21" aria-hidden="true" tabindex="-1"></a>a_n</span>
<span id="cb42-22"><a href="#cb42-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-23"><a href="#cb42-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Estimate L using Watterson's formula</span></span>
<span id="cb42-24"><a href="#cb42-24" aria-hidden="true" tabindex="-1"></a>L_est <span class="ot">&lt;-</span> <span class="fu">sum</span>(sfs[<span class="sc">-</span><span class="dv">1</span>]) <span class="sc">/</span> (<span class="dv">4</span> <span class="sc">*</span> n <span class="sc">*</span> mu <span class="sc">*</span> a_n)</span>
<span id="cb42-25"><a href="#cb42-25" aria-hidden="true" tabindex="-1"></a>L_est  <span class="co"># how close is this to the true L = 5e8?</span></span>
<span id="cb42-26"><a href="#cb42-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-27"><a href="#cb42-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Now run EPOS with L_est and compare to the true-L run</span></span>
<span id="cb42-28"><a href="#cb42-28" aria-hidden="true" tabindex="-1"></a><span class="co"># dir_epos &lt;- gl.download.binary("epos", out.dir = tempdir())</span></span>
<span id="cb42-29"><a href="#cb42-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Ne_est  &lt;- gl.run.epos(sim50, epos.path = dir_epos, L = L_est,   u = mu, boot = 10)</span></span>
<span id="cb42-30"><a href="#cb42-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Ne_true &lt;- gl.run.epos(sim50, epos.path = dir_epos, L = 5e8,     u = mu, boot = 10)</span></span>
<span id="cb42-31"><a href="#cb42-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-32"><a href="#cb42-32" aria-hidden="true" tabindex="-1"></a><span class="co"># QUESTIONS:</span></span>
<span id="cb42-33"><a href="#cb42-33" aria-hidden="true" tabindex="-1"></a><span class="co"># - How accurate is the Watterson estimate of L?</span></span>
<span id="cb42-34"><a href="#cb42-34" aria-hidden="true" tabindex="-1"></a><span class="co"># - How much does the resulting Ne trajectory differ between L_est and the true L?</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<script type="application/json" data-ui-opts="1">{"engine":"r","has_checker":false,"caption":"<span data-i18n=\"text.enginecap\" data-i18n-opts=\"{&quot;engine&quot;:&quot;R&quot;}\">R Code<\/span>"}</script>
</div>
</div>
</div>
<hr>
</section>
</section>
<section id="winding-up" class="level2">
<h2 class="anchored" data-anchor-id="winding-up">Winding Up</h2>
<section id="discussion-questions" class="level3">
<h3 class="anchored" data-anchor-id="discussion-questions">Discussion Questions</h3>
<ol type="1">
<li><p>You have a threatened frog species with <em>Nc</em> = 2000 adults and an LD <em>Ne</em> estimate of 80. What does the large <em>Nc</em>/<em>Ne</em> ratio imply about the species’ biology? What management actions might you consider?</p></li>
<li><p>A coalescent <em>Ne</em> analysis shows a wolf population had <em>Ne</em> ~ 10,000 during the Pleistocene, declining to ~500 in the last 200 years. An LD <em>Ne</em> estimate from current samples gives <em>Ne</em> = 40. Are these estimates contradictory? What is each capturing?</p></li>
<li><p>You run EPOS on two populations of the same species. Both show similar trajectory shapes but the axis values differ by a factor of three. Before concluding the populations have genuinely different histories, what alternative explanations should you rule out?</p></li>
<li><p>Why might singletons be removed (<code>minbinsize = 2</code> or higher) in empirical SFS-based analyses of real data? What biological information do you lose by removing them?</p></li>
<li><p>A colleague argues that Stairways2 is always preferable to EPOS because it uses a more sophisticated optimisation. When would you disagree, and why?</p></li>
</ol>
</section>
<section id="where-have-we-come" class="level3">
<h3 class="anchored" data-anchor-id="where-have-we-come">Where Have We Come?</h3>
<p>In this session we have covered:</p>
<ul>
<li>The definition of <em>Ne</em> and why <em>Ne</em> &lt; <em>Nc</em> in real populations — and why this matters for conservation</li>
<li>The three major classes of <em>Ne</em> estimator: drift <em>Ne</em>, LD <em>Ne</em>, and coalescent <em>Ne</em>, each measuring the same underlying concept at different timescales</li>
<li>The LD-based contemporary <em>Ne</em> using <code>gl.LDNe</code>, with emphasis on MAF filtering (<code>critical</code>), the Waples correction for genomic data, and the interpretation of <em>Inf</em> estimates</li>
<li>The Site Frequency Spectrum (SFS): what it is, how to compute it with <code>gl.sfs</code>, and what demographic signals its shape contains</li>
<li>Historical <em>Ne</em> inference using <strong>EPOS</strong> — fast and practical for exploratory work — including the critical role of <em>L</em> and <em>mu</em> in determining absolute versus relative trajectory estimates</li>
<li>The relationship between EPOS and <strong>Stairways2</strong>, and when each is appropriate</li>
</ul>
<hr>
</section>
</section>
<section id="further-reading" class="level2">
<h2 class="anchored" data-anchor-id="further-reading">Further Reading</h2>
<ul>
<li>Franklin, I. R. (1980). Evolutionary change in small populations. In: Soule, M. E. &amp; Wilcox, B. A. (eds) <em>Conservation Biology: An Evolutionary-Ecological Perspective</em>. Sinauer, Sunderland.</li>
<li>Waples, R. S. (2006). A bias correction for estimates of effective population size based on linkage disequilibrium at unlinked gene loci. <em>Conservation Genetics</em> 7: 167–184.</li>
<li>Waples, R. K. et al.&nbsp;(2016). Estimating contemporary effective population size in non-model species using linkage disequilibrium across thousands of loci. <em>Heredity</em> 117: 233–240.</li>
<li>Lynch, M. et al.&nbsp;(2019). Inference of historical population-size changes with allele-frequency data. <em>G3: Genes, Genomes, Genetics</em> 9: 2501–2512. [EPOS]</li>
<li>Liu, X. &amp; Fu, Y.-X. (2020). Stairway Plot 2: demographic history inference with folded SNP frequency spectra. <em>Genome Biology</em> 21: 280. [Stairways2]</li>
<li>Frankham, R. (1995). Effective population size/adult population size ratios in wildlife: a review. <em>Genetics Research</em> 66: 95–107.</li>
<li>Hoban, S. et al.&nbsp;(2022). Global genetic diversity status and trends: towards a suite of essential biodiversity variables (EBVs) for genetic composition. <em>Biological Reviews</em> 97: 1511–1538.</li>
</ul>
<p>
<script type="application/shiny-prerendered" data-context="server">
learnr:::register_http_handlers(session, metadata = NULL)
</script>
 
<script type="application/shiny-prerendered" data-context="server">
learnr:::prepare_tutorial_state(session)
</script>
 
<script type="application/shiny-prerendered" data-context="server">
learnr:::i18n_observe_tutorial_language(input, session)
</script>
 
<script type="application/shiny-prerendered" data-context="server">
session$onSessionEnded(function() {
        learnr:::event_trigger(session, "session_stop")
      })
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-ex1-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-ex1-code-editor`)), session)
output$`tutorial-exercise-ex1-output` <- renderUI({
  `tutorial-exercise-ex1-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
learnr:::store_exercise_cache(structure(list(label = "ex1", global_setup = structure(c("library(learnr)", 
"knitr::opts_chunk$set(echo = TRUE, eval = FALSE, cache=TRUE)"
), chunk_opts = list(label = "setup", include = FALSE)), setup = "library(dartRverse)\nlibrary(dartR.popgen)\nlibrary(ggplot2)", 
    chunks = list(list(label = "ex1-setup", code = "library(dartRverse)\nlibrary(dartR.popgen)\nlibrary(ggplot2)", 
        opts = list(label = "\"ex1-setup\"", eval = "TRUE", echo = "FALSE", 
            message = "FALSE", warning = "FALSE"), engine = "r"), 
        list(label = "ex1", code = "\ndir_ne <- gl.download.binary(\"neestimator\", out.dir = tempdir())\n\n# Simulate 50 individuals with 3000 loci\nsim50 <- gl.sim.Neconst(ninds = 50, nlocs = 3000)\n\n# Run a few generations of random mating to approach HWE\nfor (i in 1:5) {\n  sim50 <- gl.sim.offspring(sim50, sim50, noffpermother = 1)\n}\n\n\n\n\n# Estimate Ne using LD method\n# critical = c(0, 0.05): once without MAF filter, once with MAF >= 5%\nne_result <- gl.LDNe(sim50,\n                     outfile      = \"sim50LD.txt\",\n                     neest.path   = dir_ne,\n                     critical     = c(0, 0.05),\n                     singleton.rm = TRUE,\n                     mating       = \"random\")", 
            opts = list(label = "\"ex1\"", exercise = "TRUE", 
                exercise.setup = "\"ex1-setup\"", class.chunk = "\"gold-panel\"", 
                exercise.timelimit = "120"), engine = "r")), 
    code_check = NULL, error_check = NULL, check = NULL, solution = NULL, 
    tests = NULL, options = list(eval = FALSE, echo = TRUE, results = "markup", 
        tidy = FALSE, tidy.opts = NULL, collapse = FALSE, prompt = FALSE, 
        comment = NA, highlight = FALSE, size = "normalsize", 
        background = "#F7F7F7", strip.white = TRUE, cache = 3, 
        cache.path = "W07_cache/html/", cache.vars = NULL, cache.lazy = TRUE, 
        dependson = NULL, autodep = FALSE, cache.rebuild = FALSE, 
        fig.keep = "high", fig.show = "asis", fig.align = "default", 
        fig.path = "W07_files/figure-html/", dev = "png", dev.args = NULL, 
        dpi = 192, fig.ext = "png", fig.width = 7L, fig.height = 5L, 
        fig.env = "figure", fig.cap = NULL, fig.scap = NULL, 
        fig.lp = "fig:", fig.subcap = NULL, fig.pos = "", out.width = 672L, 
        out.height = NULL, out.extra = NULL, fig.retina = 2, 
        external = TRUE, sanitize = FALSE, interval = 1, aniopts = "controls,loop", 
        warning = TRUE, error = FALSE, message = TRUE, render = NULL, 
        ref.label = NULL, child = NULL, engine = "r", split = FALSE, 
        include = TRUE, purl = TRUE, fig.asp = NULL, fenced.echo = FALSE, 
        ft.shadow = FALSE, exercise.eval = FALSE, label = "ex1", 
        exercise = TRUE, exercise.setup = "ex1-setup", class.chunk = "gold-panel", 
        exercise.timelimit = 120, original.params.src = "r ex1, exercise=TRUE, exercise.setup=\"ex1-setup\", class.chunk=\"gold-panel\", exercise.timelimit = 120", 
        chunk.echo = FALSE, code = c("", "dir_ne <- gl.download.binary(\"neestimator\", out.dir = tempdir())", 
        "", "# Simulate 50 individuals with 3000 loci", "sim50 <- gl.sim.Neconst(ninds = 50, nlocs = 3000)", 
        "", "# Run a few generations of random mating to approach HWE", 
        "for (i in 1:5) {", "  sim50 <- gl.sim.offspring(sim50, sim50, noffpermother = 1)", 
        "}", "", "", "", "", "# Estimate Ne using LD method", 
        "# critical = c(0, 0.05): once without MAF filter, once with MAF >= 5%", 
        "ne_result <- gl.LDNe(sim50,", "                     outfile      = \"sim50LD.txt\",", 
        "                     neest.path   = dir_ne,", "                     critical     = c(0, 0.05),", 
        "                     singleton.rm = TRUE,", "                     mating       = \"random\")"
        ), out.width.px = 672L, out.height.px = 480L, hash = "W07_cache/html/ex1_5ab18e58fb23ecc9ea7526510f1375df", 
        params.src = "r ex1, exercise=TRUE, exercise.setup=\"ex1-setup\", class.chunk=\"gold-panel\", exercise.timelimit = 120", 
        fig.num = 0, exercise.df_print = "default", exercise.checker = "NULL"), 
    engine = "r", version = "4"), class = c("r", "tutorial_exercise"
)))
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-ex2-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-ex2-code-editor`)), session)
output$`tutorial-exercise-ex2-output` <- renderUI({
  `tutorial-exercise-ex2-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
learnr:::store_exercise_cache(structure(list(label = "ex2", global_setup = structure(c("library(learnr)", 
"knitr::opts_chunk$set(echo = TRUE, eval = FALSE, cache=TRUE)"
), chunk_opts = list(label = "setup", include = FALSE)), setup = "library(dartRverse)\nlibrary(dartR.popgen)\nlibrary(ggplot2)", 
    chunks = list(list(label = "ex1-setup", code = "library(dartRverse)\nlibrary(dartR.popgen)\nlibrary(ggplot2)", 
        opts = list(label = "\"ex1-setup\"", eval = "TRUE", echo = "FALSE", 
            message = "FALSE", warning = "FALSE"), engine = "r"), 
        list(label = "ex2", code = "# EXERCISE:\n# Simulate two populations:\n#   (a) ninds = 200, nlocs = 1000  — a large population\n#   (b) ninds = 20,  nlocs = 1000  — a small population\n#\n# For each, run 3 generations of random mating, then estimate Ne with gl.LDNe.\n# Use critical = c(0, 0.05).\n#\n# QUESTIONS:\n# - Do you get Inf estimates for the larger population?\n# - Does the MAF filter help?\n# - How close are the estimates to the true Ne?\n\ndir_ne <- gl.download.binary(\"neestimator\", out.dir = tempdir())\n\n# Large population (true Ne ~ 200)\nsim200 <- gl.sim.Neconst(ninds = 200, nlocs = 1000)\nfor (i in 1:3) sim200 <- gl.sim.offspring(sim200, sim200, noffpermother = 1)\n\n# Small population (true Ne ~ 20)\nsim20 <- gl.sim.Neconst(ninds = 20, nlocs = 1000)\nfor (i in 1:3) sim20 <- gl.sim.offspring(sim20, sim20, noffpermother = 1)\n\n# Run gl.LDNe for both and compare:\n", 
            opts = list(label = "\"ex2\"", exercise = "TRUE", 
                exercise.setup = "\"ex1-setup\"", class.chunk = "\"gold-panel\"", 
                exercise.timelimit = "120"), engine = "r")), 
    code_check = NULL, error_check = NULL, check = NULL, solution = NULL, 
    tests = NULL, options = list(eval = FALSE, echo = TRUE, results = "markup", 
        tidy = FALSE, tidy.opts = NULL, collapse = FALSE, prompt = FALSE, 
        comment = NA, highlight = FALSE, size = "normalsize", 
        background = "#F7F7F7", strip.white = TRUE, cache = 3, 
        cache.path = "W07_cache/html/", cache.vars = NULL, cache.lazy = TRUE, 
        dependson = NULL, autodep = FALSE, cache.rebuild = FALSE, 
        fig.keep = "high", fig.show = "asis", fig.align = "default", 
        fig.path = "W07_files/figure-html/", dev = "png", dev.args = NULL, 
        dpi = 192, fig.ext = "png", fig.width = 7L, fig.height = 5L, 
        fig.env = "figure", fig.cap = NULL, fig.scap = NULL, 
        fig.lp = "fig:", fig.subcap = NULL, fig.pos = "", out.width = 672L, 
        out.height = NULL, out.extra = NULL, fig.retina = 2, 
        external = TRUE, sanitize = FALSE, interval = 1, aniopts = "controls,loop", 
        warning = TRUE, error = FALSE, message = TRUE, render = NULL, 
        ref.label = NULL, child = NULL, engine = "r", split = FALSE, 
        include = TRUE, purl = TRUE, fig.asp = NULL, fenced.echo = FALSE, 
        ft.shadow = FALSE, exercise.eval = FALSE, label = "ex2", 
        exercise = TRUE, exercise.setup = "ex1-setup", class.chunk = "gold-panel", 
        exercise.timelimit = 120, original.params.src = "r ex2, exercise=TRUE, exercise.setup=\"ex1-setup\", class.chunk=\"gold-panel\", exercise.timelimit=120", 
        chunk.echo = FALSE, code = c("# EXERCISE:", "# Simulate two populations:", 
        "#   (a) ninds = 200, nlocs = 1000  — a large population", 
        "#   (b) ninds = 20,  nlocs = 1000  — a small population", 
        "#", "# For each, run 3 generations of random mating, then estimate Ne with gl.LDNe.", 
        "# Use critical = c(0, 0.05).", "#", "# QUESTIONS:", 
        "# - Do you get Inf estimates for the larger population?", 
        "# - Does the MAF filter help?", "# - How close are the estimates to the true Ne?", 
        "", "dir_ne <- gl.download.binary(\"neestimator\", out.dir = tempdir())", 
        "", "# Large population (true Ne ~ 200)", "sim200 <- gl.sim.Neconst(ninds = 200, nlocs = 1000)", 
        "for (i in 1:3) sim200 <- gl.sim.offspring(sim200, sim200, noffpermother = 1)", 
        "", "# Small population (true Ne ~ 20)", "sim20 <- gl.sim.Neconst(ninds = 20, nlocs = 1000)", 
        "for (i in 1:3) sim20 <- gl.sim.offspring(sim20, sim20, noffpermother = 1)", 
        "", "# Run gl.LDNe for both and compare:", ""), out.width.px = 672L, 
        out.height.px = 480L, hash = "W07_cache/html/ex2_1e94705ecb5c1dfb8e7f302df5239f70", 
        params.src = "r ex2, exercise=TRUE, exercise.setup=\"ex1-setup\", class.chunk=\"gold-panel\", exercise.timelimit=120", 
        fig.num = 0, exercise.df_print = "default", exercise.checker = "NULL"), 
    engine = "r", version = "4"), class = c("r", "tutorial_exercise"
)))
</script>
 
<script type="application/shiny-prerendered" data-context="server">
`tutorial-exercise-ex3-result` <- learnr:::setup_exercise_handler(reactive(req(input$`tutorial-exercise-ex3-code-editor`)), session)
output$`tutorial-exercise-ex3-output` <- renderUI({
  `tutorial-exercise-ex3-result`()
})
</script>
 
<script type="application/shiny-prerendered" data-context="server">
learnr:::store_exercise_cache(structure(list(label = "ex3", global_setup = structure(c("library(learnr)", 
"knitr::opts_chunk$set(echo = TRUE, eval = FALSE, cache=TRUE)"
), chunk_opts = list(label = "setup", include = FALSE)), setup = "library(dartRverse)\nlibrary(ggplot2)", 
    chunks = list(list(label = "ex3-setup", code = "library(dartRverse)\nlibrary(ggplot2)", 
        opts = list(label = "\"ex3-setup\"", eval = "TRUE", echo = "FALSE", 
            message = "FALSE", warning = "FALSE"), engine = "r"), 
        list(label = "ex3", code = "# EXERCISE:\n# Using an example data set \n# Steps:\n#   1. gl.filter.callrate(method = \"loc\", threshold = 0.95) to filter missing data\n#   2. gl.filter.monomorphs() to remove invariant loci\n#   3. gl.sfs(folded = TRUE) with minbinsize = 1 and then minbinsize = 5\n#\n# QUESTIONS:\n# - What does the SFS look like? Does it suggest recent expansion or contraction?\n# - How does changing minbinsize affect the number of SNPs in the spectrum?\n# - When would you use a higher minbinsize in real data?\n\nglex    <- readRDS(file.path(path.data, \"slim_ex.rds\"))\n\n", 
            opts = list(label = "\"ex3\"", exercise = "TRUE", 
                exercise.setup = "\"ex3-setup\"", class.chunk = "\"gold-panel\""), 
            engine = "r")), code_check = NULL, error_check = NULL, 
    check = NULL, solution = NULL, tests = NULL, options = list(
        eval = FALSE, echo = TRUE, results = "markup", tidy = FALSE, 
        tidy.opts = NULL, collapse = FALSE, prompt = FALSE, comment = NA, 
        highlight = FALSE, size = "normalsize", background = "#F7F7F7", 
        strip.white = TRUE, cache = 3, cache.path = "W07_cache/html/", 
        cache.vars = NULL, cache.lazy = TRUE, dependson = NULL, 
        autodep = FALSE, cache.rebuild = FALSE, fig.keep = "high", 
        fig.show = "asis", fig.align = "default", fig.path = "W07_files/figure-html/", 
        dev = "png", dev.args = NULL, dpi = 192, fig.ext = "png", 
        fig.width = 7L, fig.height = 5L, fig.env = "figure", 
        fig.cap = NULL, fig.scap = NULL, fig.lp = "fig:", fig.subcap = NULL, 
        fig.pos = "", out.width = 672L, out.height = NULL, out.extra = NULL, 
        fig.retina = 2, external = TRUE, sanitize = FALSE, interval = 1, 
        aniopts = "controls,loop", warning = TRUE, error = FALSE, 
        message = TRUE, render = NULL, ref.label = NULL, child = NULL, 
        engine = "r", split = FALSE, include = TRUE, purl = TRUE, 
        fig.asp = NULL, fenced.echo = FALSE, ft.shadow = FALSE, 
        exercise.eval = FALSE, label = "ex3", exercise = TRUE, 
        exercise.setup = "ex3-setup", class.chunk = "gold-panel", 
        original.params.src = "r ex3, exercise=TRUE, exercise.setup=\"ex3-setup\", class.chunk=\"gold-panel\"", 
        chunk.echo = FALSE, code = c("# EXERCISE:", "# Using an example data set ", 
        "# Steps:", "#   1. gl.filter.callrate(method = \"loc\", threshold = 0.95) to filter missing data", 
        "#   2. gl.filter.monomorphs() to remove invariant loci", 
        "#   3. gl.sfs(folded = TRUE) with minbinsize = 1 and then minbinsize = 5", 
        "#", "# QUESTIONS:", "# - What does the SFS look like? Does it suggest recent expansion or contraction?", 
        "# - How does changing minbinsize affect the number of SNPs in the spectrum?", 
        "# - When would you use a higher minbinsize in real data?", 
        "", "glex    <- readRDS(file.path(path.data, \"slim_ex.rds\"))", 
        "", ""), out.width.px = 672L, out.height.px = 480L, hash = "W07_cache/html/ex3_3ef375d51d713a40f4cb5f1c915662d0", 
        params.src = "r ex3, exercise=TRUE, exercise.setup=\"ex3-setup\", class.chunk=\"gold-panel\"", 
        fig.num = 0, exercise.df_print = "default", exercise.checker = "NULL"), 
    engine = "r", version = "4"), class = c("r", "tutorial_exercise"
)))
</script>
 
<script type="application/shiny-prerendered" data-context="server-extras">
ojs_define <- function(..., .session = shiny::getDefaultReactiveDomain()) {
  quos <- rlang::enquos(...)
  vars <- rlang::list2(...)
  nm <- names(vars)
  if (is.null(nm)) {
    nm <- rep_len("", length(vars))
  }
  mapply(
    function(q, nm, val) {
      # Infer name, if possible
      if (nm == "") {
        tryCatch(
          {
            nm <- rlang::as_name(q)
          },
          error = function(e) {
            code <- paste(collapse = "\n", deparse(rlang::f_rhs(q)))
            stop(
              "ojs_define() could not create a name for the argument: ",
              code
            )
          }
        )
      }
      .session$output[[nm]] <- val
      outputOptions(.session$output, nm, suspendWhenHidden = FALSE)
      .session$sendCustomMessage("ojs-export", list(name = nm))
      NULL
    },
    quos,
    nm,
    vars,
    SIMPLIFY = FALSE,
    USE.NAMES = FALSE
  )
  invisible()
}
</script>
</p>
<!--html_preserve-->
<script type="application/shiny-prerendered" data-context="dependencies">
{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboard"]},{"type":"character","attributes":{},"value":["1.7.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file","href"]}},"value":[{"type":"character","attributes":{},"value":["htmldependencies/lib/clipboard-1.7.1"]},{"type":"character","attributes":{},"value":["https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["klippy"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.0.0.9500"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["primer-tooltips"]},{"type":"character","attributes":{},"value":["1.4.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file","href"]}},"value":[{"type":"character","attributes":{},"value":["htmldependencies/lib/primer-tooltips-1.4.0"]},{"type":"character","attributes":{},"value":["https://cdn.jsdelivr.net/npm/primer-tooltips@1.4.0/build/"]}]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["build.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["klippy"]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["0.0.0.9500"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["klippy"]},{"type":"character","attributes":{},"value":["0.0.0.9500"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["htmldependencies/lib/klippy-0.0.0.9500"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["js/klippy.min.js"]},{"type":"character","attributes":{},"value":["css/klippy.min.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["klippy"]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["0.0.0.9500"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["jquery"]},{"type":"character","attributes":{},"value":["3.6.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/3.6.0"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquery-3.6.0.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquerylib"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.1.4"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["font-awesome"]},{"type":"character","attributes":{},"value":["6.5.2"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["fontawesome"]}]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["css/all.min.css","css/v4-shims.min.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["fontawesome"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.5.3"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["bootbox"]},{"type":"character","attributes":{},"value":["5.5.2"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/bootbox"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["bootbox.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.5"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["idb-keyvalue"]},{"type":"character","attributes":{},"value":["3.2.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/idb-keyval"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["idb-keyval-iife-compat.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["0.11.5"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial"]},{"type":"character","attributes":{},"value":["0.11.5"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial.js"]},{"type":"character","attributes":{},"value":["tutorial.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.5"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.5"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["2.0.10"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.5"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.5"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["2.0.10"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.5"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["ace"]},{"type":"character","attributes":{},"value":["1.10.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/ace"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["ace.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.5"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["clipboardjs"]},{"type":"character","attributes":{},"value":["2.0.10"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/clipboardjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["clipboard.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.11.5"]}]}]}
</script>
<!--/html_preserve-->
<!--html_preserve-->
<script type="application/shiny-prerendered" data-context="execution_dependencies">
{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["packages"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["packages","version"]},"class":{"type":"character","attributes":{},"value":["data.frame"]},"row.names":{"type":"integer","attributes":{},"value":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172]}},"value":[{"type":"character","attributes":{},"value":["ade4","adegenet","ape","arsenal","backports","base","base64enc","bigassertr","bigparallelr","bigsnpr","bigsparser","bigstatsr","BiocGenerics","bitops","cachem","checkmate","cli","cluster","codetools","colorspace","compiler","cowplot","crayon","dartR.base","dartR.captive","dartR.data","dartR.popgen","dartR.sexlinked","dartR.sim","dartR.spatial","dartRverse","data.table","datasets","devtools","dichromat","digest","dismo","doParallel","doRNG","dotCall64","dplyr","ellipsis","evaluate","fansi","farver","fastmap","fields","flock","fontawesome","foreach","foreign","Formula","fs","furrr","future","gdsfmt","generics","ggdendro","ggplot2","ggpmisc","ggpp","ggrepel","globals","glue","graphics","grDevices","grid","gridExtra","gtable","gtools","haplo.stats","Hmisc","htmlTable","htmltools","htmlwidgets","httpuv","httr","igraph","iterators","jquerylib","jsonlite","knitr","labeling","later","lattice","LEA","learnr","lifecycle","listenv","magrittr","maps","MASS","Matrix","MatrixModels","memoise","methods","mgcv","mime","miniUI","multcomp","mvtnorm","nlme","nnet","parallel","parallelly","patchwork","pegas","permute","pillar","pkgbuild","pkgconfig","pkgload","plyr","poisbinom","polspline","polynom","profvis","promises","purrr","quantreg","R6","raster","RColorBrewer","Rcpp","RCurl","remotes","reshape2","rlang","rmarkdown","rmio","rms","rngtools","rpart","rprojroot","rstudioapi","S7","sandwich","scales","seqinr","sessioninfo","shiny","SNPassoc","SNPRelate","snpStats","sp","spam","SparseM","splines","StAMPP","stats","stringi","stringr","survival","terra","TH.data","tibble","tidyr","tidyselect","tools","urlchecker","usethis","utf8","utils","vctrs","vegan","viridisLite","withr","xfun","xtable","yaml","zlibbioc","zoo"]},{"type":"character","attributes":{},"value":["1.7-22","2.1.10","5.8","3.6.3","1.5.0","4.4.1","0.1-3","0.1.7","0.3.2","1.12.18","0.7.3","1.6.2","0.52.0","1.0-9","1.1.0","2.3.2","3.6.3","2.1.6","0.2-20","2.1-1","4.4.1","1.1.3","1.5.3","1.2.2","1.2.2","1.2.2","1.2.2","1.2.2","1.2.2","1.2.2","1.2.2","1.16.2","4.4.1","2.4.5","2.0-0.1","0.6.37","1.3-14","1.0.17","1.8.6.2","1.2","1.1.4","0.3.2","1.0.1","1.0.6","2.1.2","1.2.0","16.3","0.7","0.5.3","1.5.2","0.8-87","1.2-5","1.6.5","0.3.1","1.34.0","1.42.0","0.1.3","0.2.0","4.0.0","0.6.1","0.5.8-1","0.9.6","0.16.3","1.7.0","4.4.1","4.4.1","4.4.1","2.3","0.3.6","3.9.5","1.9.7","5.2-0","2.4.3","0.5.8.1","1.6.4","1.6.15","1.4.7","2.1.1","1.0.14","0.1.4","1.8.9","1.49","0.4.3","1.3.2","0.22-6","3.18.0","0.11.5","1.0.4","0.9.1","2.0.3","3.4.2.1","7.3-61","1.7-1","0.5-3","2.0.1","4.4.1","1.9-1","0.12","0.1.1.1","1.4-26","1.3-1","3.1-166","7.3-19","4.4.1","1.38.0","1.3.2","1.3","0.9-7","1.9.0","1.4.5","2.0.3","1.4.0","1.8.9","1.0.1","1.1.25","1.4-1","0.4.0","1.3.0","1.0.2","5.99","2.5.1","3.6-30","1.1-3","1.0.13","1.98-1.16","2.5.0","1.4.4","1.1.4","2.29","0.4.0","6.8-2","1.5.2","4.1.23","2.0.4","0.17.1","0.2.0","3.1-1","1.4.0","4.2-36","1.2.2","1.9.1","2.1-2","1.40.0","1.56.0","2.1-4","2.11-0","1.84-2","4.4.1","1.6.3","4.4.1","1.8.4","1.5.1","3.7-0","1.7-83","1.1-2","3.2.1","1.3.1","1.2.1","4.4.1","1.0.1","3.1.0","1.2.4","4.4.1","0.6.5","2.6-8","0.4.2","3.0.2","0.49","1.8-4","2.3.10","1.52.0","1.8-12"]}]}]}
</script>
<!--/html_preserve-->


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
            // target, if specified
            link.setAttribute("target", "_blank");
            if (link.getAttribute("rel") === null) {
              link.setAttribute("rel", "noopener");
            }
            // default icon
            link.classList.add("external");
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../../../inst/tutorials/W06/W06.html" class="pagination-link" aria-label="W06 Small Populations">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">W06 Small Populations</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../../inst/tutorials/W08/W08.html" class="pagination-link" aria-label="W08 Divergence &amp; Phylogenetics">
        <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">W08 Divergence &amp; Phylogenetics</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>